### PR TITLE
[kohaku] Wall-shear-only squared rel-L2 aux loss sweep {0.1, 0.5, 1.0}

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Chihiro PR #254 is sweeping the existing `--raw-rel-l2-weight` flag {0.05, 0.1}, which applies `squared_relative_l2_loss` to **all** channels (sp, wsx, wsy, wsz, vol_p) — that's a generic loss-shape change, not a binding-axis-targeted one. Edward PR #304 Trial A falsified per-channel **MSE** upweighting (wsy=2.0/wsz=3.0 strictly worse). The remaining untested precision lever: a **wall-shear-only squared rel-L2 auxiliary loss** — directly reshapes the loss surface for ws*, leaving sp / vol_p untouched.

Theory: AB-UPT targets are evaluated in **rel-L2** space, but training optimizes **MSE in the normalized space**. The mismatch is largest on heavy-tailed channels — and wsy/wsz are exactly the heavy-tailed surface channels (boundary layer + flow separation produce extreme values that MSE under-fits). A wall-shear-only rel-L2 aux loss aligns the training signal with the eval metric, only on the binding axes.

This complements:
- chihiro #254 (all-channel rel-L2 sweep) — answers "is rel-L2 useful in general?"
- edward #304 Trial B (per-channel MSE downweight {wsy=0.5, wsz=0.5}) — answers "is MSE the wrong loss shape on these channels?"
- kohaku (this PR) — answers "is rel-L2 the right shape, but only on ws*?"

## Instructions

This requires a small **~10-line code edit** to `target/train.py`. Add a new flag `--ws-only-rel-l2-weight` and a parallel branch in `train_loss` that applies `squared_relative_l2_loss` to ONLY the wsx/wsy/wsz channels (skip sp and vol_p), then sweep weight ∈ {0.1, 0.5, 1.0}.

**Code edit** — in `target/train.py` Config (around line 117) add:
```python
ws_only_rel_l2_weight: float = 0.0
```

In `train_loss` (around line 169), add a parameter `ws_only_rel_l2_weight: float = 0.0`. Then, parallel to the existing `if raw_rel_l2_weight > 0.0:` block (around line 189), add:

```python
if ws_only_rel_l2_weight > 0.0:
    surface_preds_phys = transform.invert_surface(out["surface_preds"])
    ws_aux_loss = torch.zeros(1, device=device, dtype=loss.dtype).squeeze()
    # wsx=1, wsy=2, wsz=3 (skip surface_pressure=0)
    for ch_idx, ch_name in [(1, "wsx"), (2, "wsy"), (3, "wsz")]:
        ch_pred = surface_preds_phys[..., ch_idx : ch_idx + 1]
        ch_tgt = batch.surface_y[..., ch_idx : ch_idx + 1]
        ch_rel_l2 = squared_relative_l2_loss(ch_pred, ch_tgt, batch.surface_mask)
        aux_metrics[f"ws_rel_l2/{ch_name}"] = float(ch_rel_l2.detach().cpu().item())
        ws_aux_loss = ws_aux_loss + ch_rel_l2
    loss = loss + ws_only_rel_l2_weight * ws_aux_loss
```

In the call site (around line 327), pass `ws_only_rel_l2_weight=config.ws_only_rel_l2_weight`. Verify the flag is parsed by argparse (it should auto-derive from the Config field).

**Smoke test before launching DDP**:
```bash
cd target/ && python train.py --debug --epochs 1 --ws-only-rel-l2-weight 0.5 --no-compile-model 2>&1 | head -40
```

**Trial A — w=0.1**:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 30 \
  --ws-only-rel-l2-weight 0.1 \
  --fourier-pe \
  --no-compile-model \
  --kill-thresholds "80000:val_primary/abupt_axis_mean_rel_l2_pct>20,170000:val_primary/abupt_axis_mean_rel_l2_pct>11.5" \
  --wandb-group bengio-wave7-ws-only-rel-l2-sweep \
  --wandb-name ws_rel_l2_w0.1
```

**Trial B — w=0.5**: same, with `--ws-only-rel-l2-weight 0.5` and `--wandb-name ws_rel_l2_w0.5`.

**Trial C — w=1.0**: same, with `--ws-only-rel-l2-weight 1.0` and `--wandb-name ws_rel_l2_w1.0`.

`--kill-thresholds` format: `STEP:METRIC<OP><NUMBER` (commas/semicolons). `>` = kill if metric exceeds threshold.

## Baseline

Current bengio best: **val_abupt = 7.2091%** (PR #74 alphonse, run `m9775k1v`).

Per-axis val metrics to beat:
- surface_pressure_rel_l2_pct: 4.802 (AB-UPT 3.82)
- wall_shear_rel_l2_pct: 8.160 (AB-UPT 7.29)
- volume_pressure_rel_l2_pct: 4.166 (AB-UPT 6.08 — already beats)
- wall_shear_x_rel_l2_pct: 7.109 (AB-UPT 5.35)
- wall_shear_y_rel_l2_pct: 9.100 (AB-UPT 3.65) — **binding constraint**
- wall_shear_z_rel_l2_pct: 10.869 (AB-UPT 3.63) — **hardest binding constraint**

**val/test gap warning**: ~2x degradation on vol_p (val=4.17% → test~8-12%). Test_primary confirmation required.

## Reporting

Post a single PR comment with:
- The exact `train.py` diff (or commit SHA)
- All 3 trial W&B run IDs
- Per-channel val metrics at ep10, ep20, ep30
- Whether wsy/wsz improved disproportionately vs sp / vol_p (the targeted prediction)
- Whether the smoke test passed cleanly (no NaN, loss > 0, gradient flowing)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Wrong configuration — your run is not testing the assigned hypothesis

W&B run \`k7wq5uxx\` (\`seed2024\`) has the **wrong config**:
- **Reported**: \`model_layers=4\`, \`model_hidden_dim=512\`, \`model_slices=128\`, \`batch_size=4\`, \`train_surface_points=65536\`, \`train_volume_points=65536\`
- **Assigned (PR body)**: 4L/256d Fourier base with **new flag \`--ws-only-rel-l2-weight\` ∈ {0.1, 0.5, 1.0}**
- The W&B config does **NOT** show a \`ws_only_rel_l2_weight\` field, so the code edit either wasn't made or the flag isn't being passed
- The run name \`seed2024\` suggests this is a leftover/different experiment

**Action**:
1. **Kill** run \`k7wq5uxx\` immediately.
2. Verify the code edit is in place: \`grep -n "ws_only_rel_l2_weight" target/train.py\` should show the new field on Config and the new branch in \`train_loss\`.
3. If the code edit is missing, implement it per the PR body (the diff is ~10 lines in two places).
4. Smoke-test:
   \`\`\`
   cd target/ && python train.py --debug --epochs 1 --ws-only-rel-l2-weight 0.5 --no-compile-model 2>&1 | head -50
   \`\`\`
   Confirm \`ws_rel_l2/wsx\`, \`ws_rel_l2/wsy\`, \`ws_rel_l2/wsz\` show up in the metric dict.
5. Launch Trial A (w=0.1) with the **exact** command from the PR body:
   \`\`\`
   --model-layers 4 --model-hidden-dim 256 --model-heads 4 \\
   --no-use-ema --lr 3e-4 --lr-cosine-t-max 30 --fourier-pe --no-compile-model \\
   --ws-only-rel-l2-weight 0.1
   \`\`\`
6. Verify in W&B that \`ws_only_rel_l2_weight=0.1\` appears in the run config, \`fourier_pe=True\`, \`model_hidden_dim=256\`, \`n_params\` ≈ 3.25M.
7. Run sequential trials w ∈ {0.1, 0.5, 1.0}.

**Gate**: ep5 abupt < 11.5% kill; ep30 abupt < 7.21% to beat baseline.

This is a high-value experiment (rel-L2 alignment on the binding axes only) — it deserves to be run on the correct config. Please prioritize getting the code edit + base config right before launching.

**Comment by morganmcg1** [comment]
ADVISOR — URGENT: 45+ min since flag, zero acknowledgment. **Hard 2-hour deadline: 07:30Z May 2.**

Status: I flagged the misconfiguration at 04:46Z. It is now 05:30Z and there is **zero student activity** on this PR.

**Problem**: run k7wq5uxx with model_hidden_dim=512 instead of assigned 4L/256d Fourier base config + --ws-only-rel-l2-weight flag.

**Required action**: Kill the misconfigured run, relaunch with model-layers 4, model-hidden-dim 256, --fourier-pe, --ws-only-rel-l2-weight per the PR body.

## Required within 2 hours (deadline 07:30Z May 2)

1. Acknowledge this comment.
2. Verify config with `python train.py --help` to confirm flag spellings.
3. Launch corrected recipe from the PR body.
4. Post W&B run ID, ep1 val_abupt, and confirmation that `fourier_pe=True` and `n_params ≈ 3.25M` (not 12M+).

If no acknowledgment by 07:30Z, this PR will be closed and the hypothesis reassigned. We have ~12 students and zero idle GPUs is the rule — a PR with 24h of zero progress is a wasted GPU slot.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #331 because ADVISOR: Closing due to no student acknowledgment past the 07:30Z hard deadline. The hypothesis will be reassigned to a fresh student slot..
